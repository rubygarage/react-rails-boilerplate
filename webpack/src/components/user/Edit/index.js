import React from 'react';
import PropTypes from 'prop-types';
import Header from 'containers/Header';
import { FormattedMessage, intlShape } from 'react-intl';
import { Field } from 'redux-form';
import ErrorMessage from 'components/ui/ErrorMessage';
import AvatarUploader from 'containers/AvatarUploader';
import Button from 'components/ui/Button';

function UserEdit({
  user, handleSubmit, submitHandler,
  pristine, submitting, intl: { formatMessage },
  avatarFieldValue, avatarRemovalHandler,
}) {
  return (
    <main>
      <Header />
      <div className="jumbotron">
        <p className="lead">
          User profile edit page for {user.username}
        </p>
        <form onSubmit={handleSubmit(submitHandler)}>
          <p>{formatMessage({ id: 'user_edit.avatar' })}</p>
          {
            (user.avatarThumb && !avatarFieldValue) &&
              <div className="avatarEdit">
                <img src={user.avatarThumb} alt="User avatar" className="avatarThumb" />
                <br />
                <Button
                  className="btn btn-primary mb-3"
                  type="button"
                  disabled={!user.avatarThumb}
                  onClick={() => avatarRemovalHandler(user.id, user.avatar)}
                >
                  Remove avatar
                </Button>
              </div>
          }
          <ErrorMessage name="avatar" />
          <div className="input-group mb-3">
            <Field
              multiple={false}
              component={AvatarUploader}
              className="form-control"
              name="avatar"
            />
          </div>
          <button
            className="btn btn-primary mb-3"
            type="submit"
            disabled={pristine || submitting}
          >
            <FormattedMessage id="user_edit.save_changes" />
          </button>
        </form>
      </div>
    </main>
  );
}

UserEdit.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string,
    username: PropTypes.string,
    email: PropTypes.string,
    avatar: PropTypes.string,
    avatarThumb: PropTypes.string,
    avatarFull: PropTypes.string,
  }),
  avatarRemovalHandler: PropTypes.func.isRequired,
  intl: intlShape,
  handleSubmit: PropTypes.func.isRequired,
  submitHandler: PropTypes.func.isRequired,
  pristine: PropTypes.bool,
  submitting: PropTypes.bool,
  avatarFieldValue: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
  ]),
};

export default UserEdit;
