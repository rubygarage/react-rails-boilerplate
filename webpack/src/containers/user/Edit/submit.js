import parseJsonApiErrors from 'utils/validation';
import { SubmissionError } from 'redux-form';
import { UPDATE_USER, REQUEST } from 'constants/actions';

const submit = (values, dispatch, props) => {
  const formData = new FormData();

  Object.keys(values).forEach((key) => {
    key === 'avatar'
      ? formData.append('avatar[image]', values[key][0])
      : formData.append(key, values[key]);
  });

  return new Promise((resolve, reject) => {
    dispatch({
      type: UPDATE_USER + REQUEST, values: formData, id: props.id, resolve, reject,
    });
  }).catch((response) => {
    const errors = parseJsonApiErrors(response.errors, props.fieldsForValidation);
    throw new SubmissionError(errors);
  });
};

export default submit;
