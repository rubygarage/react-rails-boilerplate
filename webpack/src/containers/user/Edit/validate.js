const validate = (values, state) => {
  // const { formatMessage } = state.intl
  const errors = {}

  // if (!values.username) {
  //   errors.username = formatMessage({ id: 'validation.required_field' }, { field: 'Username' })
  // }
  //
  // if (!values.email) {
  //   errors.email = formatMessage({ id: 'validation.required_field' }, { field: 'Email' })
  // } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
  //   errors.email = formatMessage({ id: 'validation.invalid_email' })
  // }

  return errors
}

export default validate
