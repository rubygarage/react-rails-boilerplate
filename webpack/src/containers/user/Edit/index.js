import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getUser as getUserSaga } from 'sagas/user';
import { getUser } from 'selectors/user';
import getUserAction from 'actions/user';
import UserEditComponent from 'components/user/Edit';
import { reduxForm, formValueSelector } from 'redux-form';
import { injectIntl } from 'react-intl';
import submit from './submit';

class UserEdit extends Component {
  static propTypes = {
    user: PropTypes.shape({
      id: PropTypes.string,
      username: PropTypes.string,
      email: PropTypes.string,
      avatar: PropTypes.string,
      avatarThumb: PropTypes.string,
      avatarFull: PropTypes.string,
    }),
    id: PropTypes.string.isRequired,
    getUserAction: PropTypes.func.isRequired,
  }

  componentDidMount() {
    if (this.props.user.id) { return; }
    this.props.getUserAction(this.props.id);
  }

  avatarRemovalHandler() { // eslint-disable-line class-methods-use-this
    console.log('AVATAR REMOVAL HANDLER()');
    // TODO: delete avatar makes request to its own dedicated endpoint
  }

  render() {
    const props = {
      ...this.props,
      submitHandler: submit,
      avatarRemovalHandler: this.avatarRemovalHandler,
    };
    return <UserEditComponent {...props} />;
  }
}

const mapStateToProps = (state, ownProps) => {
  const { id } = ownProps.match.params;

  return {
    user: getUser(state, id),
    id,
    avatarFieldValue: formValueSelector('userEdit')(state, 'avatar'),
  };
};

const mapDispatchToProps = { getUserAction };

function preload(params, req, res) {
  const sagasToComplete = [];
  sagasToComplete.push([getUserSaga, params, req, res]);

  return sagasToComplete;
}
UserEdit.preload = preload;

export default connect(mapStateToProps, mapDispatchToProps)(injectIntl(reduxForm({
  form: 'userEdit',
})(UserEdit)));
