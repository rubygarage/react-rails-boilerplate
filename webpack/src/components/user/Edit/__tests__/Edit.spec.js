import React from 'react';
import { shallow } from 'enzyme';
import Edit from 'components/user/Edit';

describe('Edit user', () => {
  let props = {
    avatarRemovalHandler: jest.fn(),
    handleSubmit: jest.fn(),
    submitHandler: jest.fn(),
    intl: {
      formatMessage: (...params) => (`intl.formatMessage(${JSON.stringify(params)})`),
    },
  };

  it('shows current info at the page', () => {
    props = {
      ...props,
      user: {
        username: 'Boiler',
        email: 'example@boiler.com',
        avatar: '1',
        avatarFull: 'someaddressFull.jpg',
        avatarThumb: 'someaddressThumb.jpg',
      },
    };
    const component = shallow(<Edit {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('render update form where submit unavailable if form is pristine', () => {
    props = {
      ...props,
      pristine: true,
    };
    const component = shallow(<Edit {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('render update form where submit unavailable if form is submiting', () => {
    props = {
      ...props,
      submiting: true,
    };
    const component = shallow(<Edit {...props} />);

    expect(component).toMatchSnapshot();
  });

  it('render update form where submit available if form is dirty', () => {
    props = {
      ...props,
      dirty: true,
      pristine: false,
    };
    const component = shallow(<Edit {...props} />);

    expect(component).toMatchSnapshot();
  });
});
