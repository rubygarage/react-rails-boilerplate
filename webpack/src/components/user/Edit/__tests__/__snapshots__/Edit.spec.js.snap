// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit user render update form where submit available if form is dirty 1`] = `
<main>
  <Connect(Header) />
  <div
    className="jumbotron"
  >
    <p
      className="lead"
    >
      intl.formatMessage([{"id":"user_edit.title"},{"name":"Boiler"}])
    </p>
    <form>
      <p>
        intl.formatMessage([{"id":"user_edit.avatar"}])
      </p>
      <div
        className="avatarEdit"
      >
        <img
          alt="intl.formatMessage([{\\"id\\":\\"user_edit.avatar_alt\\"}])"
          className="avatarThumb"
          src="someaddressThumb.jpg"
        />
        <br />
        <Button
          className="btn btn-primary mb-3"
          disabled={false}
          href=""
          onClick={[Function]}
          size="md"
          to=""
          type="button"
        >
          <FormattedMessage
            id="user_edit.remove_avatar"
            values={Object {}}
          />
        </Button>
      </div>
      <ErrorMessage
        name="avatar"
      />
      <div
        className="input-group mb-3"
      >
        <Field
          className="form-control"
          component={[Function]}
          multiple={false}
          name="avatar"
        />
      </div>
      <button
        className="btn btn-primary mb-3"
        type="submit"
      >
        <FormattedMessage
          id="user_edit.save_changes"
          values={Object {}}
        />
      </button>
    </form>
  </div>
</main>
`;

exports[`Edit user render update form where submit unavailable if form is pristine 1`] = `
<main>
  <Connect(Header) />
  <div
    className="jumbotron"
  >
    <p
      className="lead"
    >
      intl.formatMessage([{"id":"user_edit.title"},{"name":"Boiler"}])
    </p>
    <form>
      <p>
        intl.formatMessage([{"id":"user_edit.avatar"}])
      </p>
      <div
        className="avatarEdit"
      >
        <img
          alt="intl.formatMessage([{\\"id\\":\\"user_edit.avatar_alt\\"}])"
          className="avatarThumb"
          src="someaddressThumb.jpg"
        />
        <br />
        <Button
          className="btn btn-primary mb-3"
          disabled={false}
          href=""
          onClick={[Function]}
          size="md"
          to=""
          type="button"
        >
          <FormattedMessage
            id="user_edit.remove_avatar"
            values={Object {}}
          />
        </Button>
      </div>
      <ErrorMessage
        name="avatar"
      />
      <div
        className="input-group mb-3"
      >
        <Field
          className="form-control"
          component={[Function]}
          multiple={false}
          name="avatar"
        />
      </div>
      <button
        className="btn btn-primary mb-3"
        disabled={true}
        type="submit"
      >
        <FormattedMessage
          id="user_edit.save_changes"
          values={Object {}}
        />
      </button>
    </form>
  </div>
</main>
`;

exports[`Edit user render update form where submit unavailable if form is submiting 1`] = `
<main>
  <Connect(Header) />
  <div
    className="jumbotron"
  >
    <p
      className="lead"
    >
      intl.formatMessage([{"id":"user_edit.title"},{"name":"Boiler"}])
    </p>
    <form>
      <p>
        intl.formatMessage([{"id":"user_edit.avatar"}])
      </p>
      <div
        className="avatarEdit"
      >
        <img
          alt="intl.formatMessage([{\\"id\\":\\"user_edit.avatar_alt\\"}])"
          className="avatarThumb"
          src="someaddressThumb.jpg"
        />
        <br />
        <Button
          className="btn btn-primary mb-3"
          disabled={false}
          href=""
          onClick={[Function]}
          size="md"
          to=""
          type="button"
        >
          <FormattedMessage
            id="user_edit.remove_avatar"
            values={Object {}}
          />
        </Button>
      </div>
      <ErrorMessage
        name="avatar"
      />
      <div
        className="input-group mb-3"
      >
        <Field
          className="form-control"
          component={[Function]}
          multiple={false}
          name="avatar"
        />
      </div>
      <button
        className="btn btn-primary mb-3"
        disabled={true}
        type="submit"
      >
        <FormattedMessage
          id="user_edit.save_changes"
          values={Object {}}
        />
      </button>
    </form>
  </div>
</main>
`;

exports[`Edit user shows current info at the page 1`] = `
<main>
  <Connect(Header) />
  <div
    className="jumbotron"
  >
    <p
      className="lead"
    >
      intl.formatMessage([{"id":"user_edit.title"},{"name":"Boiler"}])
    </p>
    <form>
      <p>
        intl.formatMessage([{"id":"user_edit.avatar"}])
      </p>
      <div
        className="avatarEdit"
      >
        <img
          alt="intl.formatMessage([{\\"id\\":\\"user_edit.avatar_alt\\"}])"
          className="avatarThumb"
          src="someaddressThumb.jpg"
        />
        <br />
        <Button
          className="btn btn-primary mb-3"
          disabled={false}
          href=""
          onClick={[Function]}
          size="md"
          to=""
          type="button"
        >
          <FormattedMessage
            id="user_edit.remove_avatar"
            values={Object {}}
          />
        </Button>
      </div>
      <ErrorMessage
        name="avatar"
      />
      <div
        className="input-group mb-3"
      >
        <Field
          className="form-control"
          component={[Function]}
          multiple={false}
          name="avatar"
        />
      </div>
      <button
        className="btn btn-primary mb-3"
        type="submit"
      >
        <FormattedMessage
          id="user_edit.save_changes"
          values={Object {}}
        />
      </button>
    </form>
  </div>
</main>
`;
